### PR TITLE
Improve GOC score calculation

### PR DESCRIPTION
**Pull request by ipilizota**
## Description

Improving the current GOC score calculation

**Related JIRA tickets:**
- ENSCOMPARASW-5668

#### Change 1
- For a gene in `neighbourhood1`, instead of searching for orthologous matches in `neighbourhood2` starting from the `previous_match+1` position onwards, search throughout the whole `neighbourhood2`.

#### Change 2
- Ignore the position of matches. Previously: A match was considered only if it was within certain distance from a previous match (previous match was a match for one of the previously searched genes from `neighbourhood1`). Now: Any match is considered, given the directionality criteria is satisfied.

## Testing
`pytest`

### CODE REVIEW

File: scripts/pipeline/orthology_benchmark.py [hunk 1]
```diff
@@ -544,47 +542,41 @@ def calculate_goc_genes(gene1: str, gene2: str, neighbourhood1: pandas.DataFrame
...
-                #   shared_orthologs++
-                curr_pos_j = index
-                if prev_match < 0 or abs(curr_pos_j - prev_match) <= allowed_gap + 1:
-                    shared_orthologs += 1
-
-                prev_match = curr_pos_j
+                shared_orthologs += 1
                 break  # Don't look for multiple matches
 
     return (shared_orthologs / (2 * n)) * 100
```
Comment (ipilizota) [inline scripts/pipeline/orthology_benchmark.py:582]: We were talking about keeping it simple and dividing by `2*n` which is the maximum number of neighbours fetched for `gene1`. But `gene1` can have fewer neighbours due to the fragmented assembly or simply by being the first or last one on the chromosome. Based on my code it is easy to divide by the number of neighbours we actually have for `gene1` but that has its pros and cons.

File: scripts/pipeline/orthology_benchmark.py [hunk 2]
```diff
@@ -544,47 +542,41 @@ def calculate_goc_genes(gene1: str, gene2: str, neighbourhood1: pandas.DataFrame
     strand2 = neighbourhood2.iloc[index2]["strand"]
```
Comment (thiagogenez) [inline scripts/pipeline/orthology_benchmark.py:542]: There is straight access to the data frame array before checking the existence of indexes in the array. Would a safety sanity check need here?e

File: src/python/tests/test_orthology_benchmark.py [hunk 3]
```diff
@@ -515,44 +515,37 @@ def test_get_neighbours(gene: str, n: int, exp_genes: List[Tuple[str, str, float
...
         ("36", "22",
          [["3", "34", 301, "+"], ["3", "35", 401, "+"], ["3", "36", 501, "+"], ["3", "37", 601, "+"],
           ["3", "38", 701, "+"]],
          [["2", "21", 1, "+"], ["2", "22", 101, "+"], ["2", "23", 201, "+"], ["2", "24", 301, "+"],
           ["2", "25", 401, "+"], ["2", "27", 601, "+"], ["2", "28", 701, "+"], ["2", "29", 801, "+"]],
-         2, 4, 50)
+         2, 100)
     ]
 )
 def test_calculate_goc_genes(gene1: str, gene2: str, neighbours1: List[Tuple[str, str, int, str]],
```
Comment (thiagogenez) [inline src/python/tests/test_orthology_benchmark.py:547]: Does the test only assume valid inputs (with valid data and invalid data)?  If affirmative, should be interesting to include invalid inputs as well (i.e. empty data frame or dataframe with negative values)?

**Review by thiagogenez — COMMENTED** [review]
Make a few comments regarding the code

File: scripts/pipeline/orthology_benchmark.py [hunk 4]
```diff
@@ -544,47 +542,41 @@ def calculate_goc_genes(gene1: str, gene2: str, neighbourhood1: pandas.DataFrame
...
 
     if strand1 != strand2:
+        # Reverse `neighbourhood2` to potentially speed up search for putative orthologous matches later
         neighbourhood2 = neighbourhood2.reindex(index=neighbourhood2.index[::-1])
         neighbourhood2.reset_index(inplace=True)
         neighbourhood2.drop("index", 1, inplace=True)
+        index2 = neighbourhood2.index[neighbourhood2["gene_id"] == gene2][0]
         strand_mismatch = True
     else:
         strand_mismatch = False
```
[the remark this answers is not in the document]
Comment (dthybert) [reply scripts/pipeline/orthology_benchmark.py:553]: yes I agree it will be more readable and the code easier to extend if required in the future

**Review by dthybert — APPROVED** [review]
I approve with the assumption that you will take Thiago comment into account before merging

File: scripts/pipeline/orthology_benchmark.py (hunk 1 above)
Comment (ipilizota) [reply scripts/pipeline/orthology_benchmark.py:582]: @dthybert Not sure if "👍" means to keep it as it is or divide by the actual number of neighbours?

File: scripts/pipeline/orthology_benchmark.py (hunk 2 above)
Comment (ipilizota) [reply scripts/pipeline/orthology_benchmark.py:542]: You're right. I should add safety sanity checks (maybe not just at this place) and then test cases accordingly. That will also address your comment in the test script 🙂

File: src/python/tests/test_orthology_benchmark.py (hunk 3 above)
Comment (ipilizota) [reply src/python/tests/test_orthology_benchmark.py:547]: Well spotted 🙂 As I said above in one of my comments, I will address this. Thank you!

File: scripts/pipeline/orthology_benchmark.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (ipilizota) [reply scripts/pipeline/orthology_benchmark.py:553]: Will do 🙂 Thank you 🙂

**Comment by ipilizota** [comment]
@thiagogenez 

> try to run
> 
> ```
> git cherry-pick 95c1fc8c6124ca82e4ee33ba140b54f4b3d45039
> ```
> 
> to solve the `pylint` problem

```
fatal: bad object 95c1fc8c6124ca82e4ee33ba140b54f4b3d45039
```
I'll get back to it tomorrow and find a way. Thank you anyways 🙂

**Comment by thiagogenez** [comment]
Travis `Pylint` job is failing on all branches. I need to check it out.

**Comment by thiagogenez** [comment]
Regarding Travis `Perl external unit tests on the latest version, with code coverage` you need to either rebase your branch with `main` or edit the file `travisci/perl-external_unittest_harness.sh` similar to @CristiGuijarro 's [commit](https://github.com/Ensembl/ensembl-compara/commit/74402b87cdc2b25b95d42c611680aa2849aa3094)

**Comment by ipilizota** [comment]
Thank you @thiagogenez. Yeah, I have to rebase it against `main` and I expect that will sort out the Perl side.
`pylint` 😭

**Review by thiagogenez — APPROVED** [review]
Please just read my comments in the main thread conversation.

**Comment by ipilizota** [comment]
> Please just read my comments in the main thread conversation.

You mean that I should ignore your suggestions in the code? 🙊 I have already implemented safety sanity checks in the script (haven't pushed them yet) and I was planning to add some test cases for them. I see you've already approved this PR. Do you mind if I ping you to have a look at this PR again once I push everything regarding safety sanity checks?

**Comment by ipilizota** [comment]
@thiagogenez 🙊
So... I included safety sanity checks and then `pylint` complained: `R0912: Too many branches (13/12) (too-many-branches)`. So I refactored the code, i.e. added one more function and corresponding test. The only suggestion that you made and I didn't incorporate was to use negative start positions for genes. That info is not used in this function at all.

I also had to edit a bit some old tests because `pylint` started complaining about them.

**Review by thiagogenez — APPROVED** [review]
Recommended improvement suggestions, such as extra sanity checks and tests, have been nicely addressed and implemented.

**Comment by ipilizota** [comment]
Travis is failing on `pylint` - being addressed in https://github.com/Ensembl/ensembl-compara/pull/482 & will be pulled on the project branch upon PR approval.

**Pull request merged**